hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/test/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/launcher/TestContainerLaunchParameterized.java

```diff
@@ -0,0 +1,148 @@
+        Arguments.of("${parameter%word}", asSet("parameter")),
+        Arguments.of("${parameter/pattern/string}", asSet("parameter")),
+        Arguments.of("${parameter^pattern}", asSet("parameter")),
+        Arguments.of("${parameter^^pattern}", asSet("parameter")),
+        Arguments.of("${parameter,pattern}", asSet("parameter")),
+        Arguments.of("${parameter,,pattern}", asSet("parameter")),
+        Arguments.of("${parameter@o}", asSet("parameter")),
+        Arguments.of("handle '${A}' simple quotes", asSet("A")),
+        Arguments.of("handle '${A} $B ${C:-$D}' simple quotes",
+            asSet("A", "B", "C", "D")),
+        Arguments.of("handle \"'${A}'\" double and single quotes", asSet()),
+        Arguments.of("handle \"'\\${A}'\" double and single quotes", asSet()),
+        Arguments.of("handle '\\${A} \\$B \\${C:-D}' single quotes", asSet()),
+        Arguments.of("handle \"${A}\" double quotes", asSet("A")),
+        Arguments.of("handle \"${A} $B ${C:-$D}\" double quotes",
+            asSet("A", "B", "C", "D"))
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputForGetEnvDependenciesLinux")
+  void testGetEnvDependenciesLinux(String input,
+                                   Set<String> expected) {
+    ContainerLaunch.ShellScriptBuilder win =
+        ContainerLaunch.ShellScriptBuilder.create(Shell.OSType.OS_TYPE_LINUX);
+    Assert.assertEquals("Failed to parse " + input, expected,
+        win.getEnvDependencies(input));
+  }
+
+  private static Stream<Arguments> inputForGetEnvDependenciesWin() {
+    return Stream.of(
+        Arguments.of(null, asSet()),
+        Arguments.of("", asSet()),
+        Arguments.of("A", asSet()),
+        Arguments.of("%%%%%%", asSet()),
+        Arguments.of("%%A%", asSet()),
+        Arguments.of("%A", asSet()),
+        Arguments.of("%A:", asSet()),
+        Arguments.of("%A%", asSet("A")),
+        Arguments.of("%%%A%", asSet("A")),
+        Arguments.of("%%C%A%", asSet("A")),
+        Arguments.of("%A:~-1%", asSet("A")),
+        Arguments.of("%A:%", asSet("A")),
+        Arguments.of("%A:whatever:a:b:%", asSet("A")),
+        Arguments.of("%A%B%", asSet("A")),
+        Arguments.of("%A%%%%%B%", asSet("A")),
+        Arguments.of("%A%%B%", asSet("A", "B")),
+        Arguments.of("%A%%%%B%", asSet("A", "B")),
+        Arguments.of("%A%:%B%:%C%:pathlist var", asSet("A", "B", "C")),
+        Arguments.of("%A%\\\\foo\\\\bar:%B%:%C%:pathlist var",
+            asSet("A", "B", "C"))
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputForGetEnvDependenciesWin")
+  void testGetEnvDependenciesWin(String input,
+                                 Set<String> expected) {
+    ContainerLaunch.ShellScriptBuilder win =
+        ContainerLaunch.ShellScriptBuilder.create(Shell.OSType.OS_TYPE_WIN);
+    Assert.assertEquals("Failed to parse " + input, expected,
+        win.getEnvDependencies(input));
+  }
+
+  private static Set<String> asSet(String... str) {
```

Comment:
Nit: Guava has the same functionality as Sets.newHashSet.

Reply:
hadoop-yarn-server-nodemanager does not depend on Guava yet

Reply:
But org.apache.hadoop.util.Sets should contain a method like this.

Reply:
Replaced with the hadoop.utils.Sets alternative.